### PR TITLE
fix: keep optional params coming from a parent record

### PR DESCRIPTION
**Pull request by chuyuandu**
resolve TODO: only keep optional params coming from a parent record

### CODE REVIEW

File: packages/router/src/matcher/index.ts [hunk 1]
```diff
@@ -277,8 +277,15 @@ export function createRouterMatcher(
         paramsFromLocation(
           currentLocation.params,
           // only keep params that exist in the resolved location
-          // TODO: only keep optional params coming from a parent record
-          matcher.keys.filter(k => !k.optional).map(k => k.name)
+          // only keep optional params coming from a parent record
+          matcher.keys
+            .filter(k => !k.optional)
+            .map(k => k.name)
```
Comment (posva) [inline packages/router/src/matcher/index.ts:283]: This could be done once at the end

**Review by posva — COMMENTED** [review]
Thanks!

- It should also go for ancestors, not just direct parents.
- It should have tests about how params with matching names but coming from different routes are discarded

But I'm actually unsure if this is the desired behavior. Currently, matching params are silently passed around and people might find that useful. I wonder if making this behavior stricter will be better in the long term

File: packages/router/src/matcher/index.ts (hunk 1 above)
Comment (chuyuandu) [reply packages/router/src/matcher/index.ts:283]: 1.  A direct parent's keys contains all optional params, includes ancestors
2. I'll open a new PR to improve the code if the behavior it is necessary, absolutely i think this is better to keep the optional params from parent record.

```js
{
    path: '/:type?/root',
    children: [
    {
        path: 'path_a',
        name: 'a',
    },
    {
        path: 'path_b',
        name: 'b',
    }
  ]
}

// current path:   '/optionType/root/path_a'

resolve({
    name: 'b'
});
```
I think `/optionType/root/path_b` is better than `/root/path_b`  in this case

Comment (posva) [reply packages/router/src/matcher/index.ts:283]: I guess o forgot how I wrote my own code 😂

Comment (chuyuandu) [reply packages/router/src/matcher/index.ts:283]: I suggest to keep optional params coming from a parent record,  that is really useful

**Comment by chuyuandu** [comment]
@posva  hey, i have improved the code and add unit tests

**Comment by posva** [comment]
thanks! I will check later on this year

**Review by posva — APPROVED** [review]

**Pull request merged**